### PR TITLE
New Python Boilerplate code added for Data Catalog Tagging using Tag Templates

### PR DESCRIPTION
**Pull request by gdmegha**
added 4 python code files which are boilderplate code for Catalog tagging using tag templates for BQ tables and columns

It handles , creation , updation and deletion of tags that were created using Data catalog tag templates.

Added DDL sample sql for the BQ master table for storing all tag related information

Added Sample JSON file which will be input and can be changed on the basis of tag template created by the user

Readme file attached

This process is reviewed as part of asset creation for for Data Governance for kithub by Pradeep Bhattiprolu. clonable code will be commited in PSO github repository , readme file will have high level overview &steps and the overall documenation will be added to kithub repository.

### CODE REVIEW

**Comment by CYarros10** [comment]
Hi! thanks for your pull request - Can you please run [pylint](https://pylint.readthedocs.io/en/stable/) and [black](https://github.com/psf/black) formatting on the python files you've added?

**Comment by gdmegha** [comment]
hi @CYarros10 , Thank you for your feedback . I have formatted using Black and pylint. Can you please review again. Thank you.

**Comment by CYarros10** [comment]
@gdmegha can you please correct the code based on the other comments I suggested as well?

**Comment by gdmegha** [comment]
hi @CYarros10 , I have corrected code based on pylint and black . Please review again

**Comment by agold-rh** [comment]
@CYarros10 Are you good with the corrections? Inquiring minds want to know :)

**Comment by CYarros10** [comment]
Hi @agold-rh - waiting for @gdmegha to address the in-line comments I've made

File: tools/datacatalog-tagging-using-tag-templates/tagging/attach_delete_tag_to_table_column.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+       sys.exit(1)
+
+
+  
+  
+      
+if __name__ == '__main__':
+  
+  
+   specificColumnTaggingList,allColumnTaggingList , tableTaggingList = attach_config("test-datahub.test1.Enterprise_Data_Catalog_Master")
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/attach_delete_tag_to_table_column.py]: changed code to get the table details from argument

File: tools/datacatalog-tagging-using-tag-templates/tagging/insert_landing.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
+"""
+Copyright 2023 Google LLC
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/insert_landing.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+datacatalog = datacatalog_v1.DataCatalogClient()
+from google.cloud import bigquery
+import argparse
+import re
+
+import sys
+import multiprocessing
+
+
+client = datacatalog_v1.DataCatalogClient()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: corrected it

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+def end_program_execution():
+ print("Program execution finished.")
+ sys.exit(1)
+
+
+def search_entries(projects,search_string,level,mode,logs_enabled=False):
+  
+  
+   level='table' if level == 'column' else level
+   #datacatalog = datacatalog_v1.DataCatalogClient()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+
+ Returns:
+   A list of dicts containing the results of the lookup. Each dict contains the
+   following keys:
+     - entry_name: The name of the entry.
+     - tag_template_name: The name of the tag template.
+     - tag_value: The value of the tag.
+ """
+ client = datacatalog_v1.DataCatalogClient()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+ result=client.list_tags(parent=entry)
+ for response in result :
+   if response.template == tag.template and response.column == tag.column :
+       tag=response
+       #tag_exists=True
+
+
+ tags_list=tag_attributes.split(",")
+ #if len(tag_attributes) > 1 else tag
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done, removed all the comments

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+           tag.fields[field_id].double_value = field_value
+       elif created_tags[field_id] == datacatalog_v1.types.FieldType.PrimitiveType.TIMESTAMP:
+           tag.fields[field_id] = datacatalog_v1.types.TagField()
+           tag.fields[field_id].timestamp_value = field_value
+       elif created_tags[field_id] == datacatalog_v1.types.FieldType.PrimitiveType.RICHTEXT:
+           tag.fields[field_id] = datacatalog_v1.types.TagField()
+           tag.fields[field_id].richtext_value = field_value
+      
+ 
+  #update_mask = FieldMask(paths=["fields"])
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+ #print(update_mask)
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+ except Exception as exception:
+   print(f"Exception Name: {type(exception).__name__} Exception Message: {exception}")
+   return exception
+
+
+
+
+if __name__ == '__main__':
+   #codeConfig = read_config_json()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/tag_util.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/upsert_staging.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+"""
+
+from google.cloud import bigquery
+
+
+# Construct a BigQuery client object.
+client = bigquery.Client()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/upsert_staging.py]: done

File: tools/datacatalog-tagging-using-tag-templates/tagging/attach_delete_tag_to_table_column.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+   self.tag_template = tag_template
+   self.tag_fields = tag_fields
+   self.tag = tag
+
+
+def delete_tag(config_table):
+   # Opening  table
+   try:
+       catalogclient = datacatalog_v1.DataCatalogClient()
+       client = bigquery.Client()
```
[the remark this answers is not in the document]
Comment (gdmegha) [reply tools/datacatalog-tagging-using-tag-templates/tagging/attach_delete_tag_to_table_column.py]: done

**Comment by gdmegha** [comment]
hi @CYarros10 all inline comments are resolved , mainly
* licence year 2023 to 2024
* clients are declared with correct prefix to avoid confusion
* parameterised - user to pass the projectname , dataset name and table name wherever required as arguments
* removed unnecesary comments

Please review

**Review by CYarros10 — APPROVED** [review]
LGTM

**Review by agold-rh — APPROVED** [review]

**Comment by agold-rh** [comment]
@gdmegha Could you please update this PR from main so that I can merge?

**Pull request merged**